File: go/easy/1_two_sum.go
// 1. Two Sum
//
// Given an array of integers nums and an integer target, return indices of the
// two numbers such that they add up to target.
//
// You may assume that each input would have exactly one solution, and you may
// not use the same element twice.
//
// You can return the answer in any order.
package main

func twoSum(nums []int, target int) []int {
	m := populateMap(nums)

	for i, x := range nums {
		y := target - x
		j, ok := m[y]
		if i != j && ok {
			return []int{i, j}
		}
	}

	return nil
}

func populateMap(nums []int) map[int]int {
	res := map[int]int{}
	for i, n := range nums {
		res[n] = i
	}

	return res
}

File: python/easy/1_two_sum.py
"""
1. Two Sum

Given an array of integers nums and an integer target, return indices of the
two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may
not use the same element twice.

You can return the answer in any order.
"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = populate_dict(nums)

        for x in nums:
            y = target - x

            if x == y and nums.count(x) > 1:
                f = nums.index(x)
                s = nums.index(x, f+1)
                return f, s

            if x != y and y in d:
                return d[x], d[y]

def populate_dict(nums):
    d = {}
    for i, num in enumerate(nums):
        d[num] = i
    return d
